src/libsync/propagateuploadng.cpp

```diff
@@ -0,0 +1,401 @@
+}
+
+/*
+  State machine:
+
+     *----> doStartUpload()
+            Check the db: is there an entry?
+              /               \
+             no                yes
+            /                   \
+           /                  PROPFIND
+       startNewUpload() <-+        +----------------------------\
+          |               |        |                             \
+         MKCOL            + slotPropfindFinishedWithError()     slotPropfindFinished()
+          |                                                       |
+      slotMkColFinished()                                         |
+          |                                                       |
+    +-----+-------------------------------------------------------+
+    |
+    +---->  startNextChunk()  ---finished?  --+
+                  ^               |          |
+                  +---------------+          |
+                                             |
+    +----------------------------------------+
+    |
+    +-> MOVE ------> moveJobFinished() ---> finalize()
+
+
+ */
+
+void PropagateUploadFileNG::doStartUpload()
+{
+    _duration.start();
+    _propagator->_activeJobList.append(this);
+
+    const SyncJournalDb::UploadInfo progressInfo = _propagator->_journal->getUploadInfo(_item->_file);
+    if (progressInfo._valid && Utility::qDateTimeToTime_t(progressInfo._modtime) == _item->_modtime ) {
+        _transferId = progressInfo._transferid;
+        auto url = chunkUrl();
+        auto job = new LsColJob(_propagator->account(), url, this);
+        _jobs.append(job);
+        job->setProperties(QList<QByteArray>() << "resourcetype" << "getcontentlength");
+        connect(job, SIGNAL(finishedWithoutError()), this, SLOT(slotPropfindFinished()));
+        connect(job, SIGNAL(finishedWithError(QNetworkReply*)),
+                this, SLOT(slotPropfindFinishedWithError()));
+        connect(job, SIGNAL(destroyed(QObject*)), this, SLOT(slotJobDestroyed(QObject*)));
+        //TODO: port to Qt4
+        connect(job, SIGNAL(directoryListingIterated(QString,QMap<QString,QString>)),
+                this, SLOT(slotPropfindIterate(QString,QMap<QString,QString>)));
+        job->start();
+        return;
+    }
+
+    startNewUpload();
+}
+
+void PropagateUploadFileNG::slotPropfindIterate(const QString &name, const QMap<QString,QString> &properties)
+{
+    if (name == chunkUrl().path()) {
+        return; // skip the info about the path itself
+    }
+    bool ok = false;
+    auto chunkId = name.mid(name.lastIndexOf('/')+1).toUInt(&ok);
+    if (ok) {
+        this->_serverChunks[chunkId] = properties["getcontentlength"].toULongLong();
+    }
+}
+
+void PropagateUploadFileNG::slotPropfindFinished()
+{
+    auto job = qobject_cast<LsColJob *>(sender());
+    slotJobDestroyed(job); // remove it from the _jobs list
+    _propagator->_activeJobList.removeOne(this);
+
+    _currentChunk = 0;
+    _sent = 0;
+    while (_serverChunks.contains(_currentChunk)) {
```

Comment:
But with the lscol job info from before we should exactly know which parts are already on the server. But for that, the upload can only be started after the lscol job info is parsed. Not sure if that is the case with the current code flow.